Documentation related to reauthentication (#217)
glpi-project/docdev: source/devapi/reauthentication.rst

                  ├─ stores the requested URL, HTTP method and POST/GET data in the session
                  ├─ stores the origin URL (referer) for the "Cancel" button
                  └─ throws RedirectException → /ReAuth/Prompt

   2. GET /ReAuth/Prompt
        └─ renders pages/reauth/prompt.html.twig
             (label + strategy template + Verify/Cancel buttons,
              form action = strategy getVerifyUrl() / getVerifyHttpMethod())

   3. POST /ReAuth/Verify
        ├─ ReAuthManager::verify($request) → strategy verify()
        ├─ on failure: the prompt is displayed again with an "Authentication failure" alert
        └─ on success:
             ├─ ReAuthManager::authenticate() → opens the 15 min window
             └─ renders pages/redirect_post.html.twig, which replays the initial request
                (same URL, same method, same POST data)

The replay is what makes the detour transparent: a submitted form is not lost, the user lands
on the page they asked for.

Session keys used
+++++++++++++++++

``glpi_reauth_until`` (expiration timestamp), ``glpi_reauth_requested_url``,
``glpi_reauth_requested_httpmethod``, ``glpi_reauth_requested_post_data``,

> ``glpi_reauth_origin_url``.

Reviewer: The session keys are listed as a run-on sentence; every other multi-item enumeration in this page (native strategies, protected itemtypes) uses a table or list. A `list-table` would be easier to scan here too.

```suggestion
.. list-table::
   :header-rows: 1
   :widths: 40 60

   * - Session key
     - Holds
   * - ``glpi_reauth_until``
     - Expiration timestamp of the current window.
   * - ``glpi_reauth_requested_url``
     - The URL to replay once verified.
   * - ``glpi_reauth_requested_httpmethod``
     - The HTTP method of the replayed request.
   * - ``glpi_reauth_requested_post_data``
     - The POST data of the replayed request.
   * - ``glpi_reauth_origin_url``
     - The referer, used for the "Cancel" button.
```

Author: fixed in 139f2f583d357ead389cf610ee2ef6d5e501b59f

check by an item check. This is exactly what was done for the plugins and marketplace pages:

.. code-block:: diff

   - Session::checkRight("config", UPDATE);
   + (new Config())->checkGlobal(UPDATE);

.. warning::

   ``Session::checkRight()`` and friends silently bypass re-authentication. If a sensitive page
   keeps using them, it stays unprotected even though its itemtype declares
   ``itemTypeRequiresReauthentication()``.

The ``$reauth_needed`` by-reference parameter of ``can()`` / ``canGlobal()`` exists for the
places that must distinguish *"forbidden"* from *"allowed, but a fresh proof of identity is
needed"* — typically to keep displaying a form whose submission will prompt later. Only trust
it when the method returned ``false``.

Step 3 — pages that are not itemtype-based
++++++++++++++++++++++++++++++++++++++++++

When the protected action does not map to an itemtype at all, call the manager directly:

.. code-block:: php

   <?php

   use Glpi\Security\ReAuth\ReAuthManager;

   // legacy front file (front/updatepassword.php)
   ReAuthManager::getInstance()->checkReAuthenticationOrRedirect();

.. code-block:: php

   <?php

> 

Reviewer: Maybe add `use` statements?
```suggestion

   use Glpi\Controller\AbstractController;
   use Glpi\Exception\Http\AccessDeniedHttpException;
   use Glpi\Security\ReAuth\ReAuthManager;
   use Symfony\Component\HttpFoundation\Response;
   use Symfony\Component\Routing\Attribute\Route;

```

Author: no, let's focus on code related to the feature.

     // No redirection is possible here: the file is called through AJAX.
     $reauth_needed = null;
     $allowed = $item->canGlobal(UPDATE, $reauth_needed);
     if (!$allowed && !$reauth_needed) {
         throw new AccessDeniedHttpException('Missing authorization');
     }

* ``processMassiveActionsForOneItemtype()`` needs no change: it is only reached after the
  prompt has been passed, so ``$item->can()`` behaves as usual.

Providing a re-authentication strategy from a plugin
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A plugin can contribute its own verification method — typically an OAuth/SSO plugin verifying
the identity through the identity provider. Registered strategies are merged with the native
ones and take part in the same priority-based selection: ``ReAuthManager::registerStrategy()``
is where it starts.

A plugin supplies only the *how* of verifying the identity. It cannot change the window
duration, nor which itemtypes are sensitive.

See :doc:`/plugins/reauthentication` for the whole recipe: registering the strategy,
implementing an in-place one, providing the prompt template, and delegating the verification to

> an external service.

Reviewer: The "how a plugin registers a strategy" paragraph duplicates content already explained in `plugins/reauthentication.rst:27-28` (registered strategies merged with native ones, same priority-based selection). The core page only needs the entry point and the pointer to the full recipe.

```suggestion
A plugin can contribute its own verification method (typically an OAuth/SSO plugin verifying
the identity through the identity provider) via ``ReAuthManager::registerStrategy()``. It
supplies only the *how* of verifying the identity: it cannot change the window duration, nor
which itemtypes are sensitive.

See :doc:`/plugins/reauthentication` for the whole recipe: registering the strategy,
implementing an in-place one, providing the prompt template, and delegating the verification to
an external service.

Re-authentication ("sudo mode")
-------------------------------

.. versionadded:: 12.0

   Re-authentication is only available from GLPI 12.0.

>    Nothing described on this page exists in GLPI 11 or earlier.

Reviewer 2: ```suggestion
```

Goals
^^^^^

Being authenticated is not enough to reach the most sensitive parts of GLPI (users, profiles
and rights, authentication settings, plugins, configuration, logs…). Before such a page is
displayed or such an action is performed, GLPI asks the user to prove their identity one more
time. This extra step is called **re-authentication**, informally *sudo mode*, by analogy with
the ``sudo`` command.

> It mitigates the misuse of an already opened session: unattended workstation, stolen or

Reviewer 3: The only thing this effectively helps protect against is an unattended workstation IMO. The rest are just a coincidence.

Author: Not only, I'm rewriting this part...

Author: done, see last commit.

replayed session cookie, forged link (CSRF-like), injected script. None of these carry the
user's password, TOTP code or SSO credentials.

> Once a verification succeeds, the user gets a **15 minutes** window

Reviewer 3: ```suggestion
Once a verification succeeds, the user gets a **15 minute** window
```

Reviewer 3: The "minute" unit functions as an adjective here instead of a noun.
Same case as the phrase "I have 2 apple drinks" where "apple" is normally a noun but is used to describe the drinks instead.

Protection is declared **per itemtype**, and enforced by the usual right check methods. There
is no per-page flag and no route attribute.

Step 1 — declare the itemtype as sensitive
++++++++++++++++++++++++++++++++++++++++++

Override ``itemTypeRequiresReauthentication()`` (defined on ``CommonGLPI``, returns ``false``
by default):

.. code-block:: php

   <?php

   use Override;

   class MySensitiveItem extends CommonDBTM
   {
       #[Override]
       protected static function itemTypeRequiresReauthentication(): bool
       {
           return true;
       }
   }

Core examples: ``User``, ``Profile``, ``Profile_User``, ``Group``, ``Group_User``, ``Config``,
``AuthLDAP``, ``AuthMail``, ``OAuthClient``, ``Glpi\Event``, ``Glpi\Inventory\Conf``.

The derived state is read through the ``final`` method
``CommonGLPI::isUserReauthenticationNeeded()``, which returns ``true`` only when the itemtype
requires it, the context is an HTTP request (not API, not CLI), and the current session has no
valid window.

Step 2 — use a right check that enforces it
+++++++++++++++++++++++++++++++++++++++++++

Nothing else is needed **if** the page checks its rights with the standard methods, because

> they already handle the redirection. The declaration is not right-specific: once an itemtype is

Reviewer 3: ```suggestion
they already handle the redirection. The declaration is not right-specific; once an itemtype is
```

sensitive, ``READ`` is prompted just like ``UPDATE``, so simply displaying such a page requires
a fresh proof of identity.

.. list-table::
   :header-rows: 1
   :widths: 40 60

   * - Method
     - Behaviour when re-authentication is missing
   * - ``$item->check($id, $right, $input)``
     - Redirects to the prompt (throws ``RedirectException``).
   * - ``$item->checkGlobal($right)``
     - Redirects to the prompt.
   * - ``$itemtype::checkReAuthenticationOrRedirect()``
     - Redirects to the prompt. Use it when there is no item to check (a page displaying a
       list, a tool page…). It checks **no right at all**: keep your own right check.
   * - ``$item->can($id, $right, $input, $reauth_needed)``
     - Returns ``false`` and sets ``$reauth_needed`` to ``true``. **Does not** redirect.
   * - ``$item->canGlobal($right, $reauth_needed)``
     - Same as ``can()``.
   * - ``Session::haveRight()`` / ``Session::checkRight()``
     - **Nothing.** These are module-level right checks; they know nothing about
       re-authentication.

So the most common change when protecting an existing legacy page is to replace a bare right

> check by an item check. This is exactly what was done for the plugins and marketplace pages:

Reviewer 3: ```suggestion
So, the most common change when protecting an existing legacy page is to replace a plain right
check by an item check. This is what was done for the plugin and marketplace pages:
```

.. code-block:: diff

   - Session::checkRight("config", UPDATE);
   + (new Config())->checkGlobal(UPDATE);

.. warning::

>    ``Session::checkRight()`` and friends silently bypass re-authentication. If a sensitive page

Reviewer 3: ```suggestion
   ``Session::checkRight()`` and similar functions silently bypass re-authentication. If a sensitive page
```

   both in the right order: calling ``checkReAuthenticationOrRedirect()`` on top of them is
   redundant.

Generic controllers
+++++++++++++++++++

``GenericFormController`` and ``GenericListController`` already call
``$class::checkReAuthenticationOrRedirect()``. An itemtype served by them is protected as soon
as it declares ``itemTypeRequiresReauthentication()``.

AJAX and non-HTML endpoints
+++++++++++++++++++++++++++

An AJAX request cannot display the prompt, so ``redirectToReauth()`` throws an
``AccessDeniedHttpException`` instead of redirecting (the check is
``!$request->isXmlHttpRequest() && $request->getPreferredFormat() === 'html'``).

Consequences when designing a sensitive feature:

* Do not put a sensitive action behind an AJAX-only endpoint, or the user will get a plain
  "access denied" with no way to recover other than reloading the page.
* Keep the *sensitive* part on a full page request, and let AJAX handle the non-sensitive
  parts.
* A sub-form loaded by AJAX may be displayed even though re-authentication is missing (see
  massive actions below), as long as the actual processing goes through a full page request

>   that will prompt.

Reviewer 3: Are we saying that this security feature effectively kills any migration towards a more SPA architecture and kills interactive feature islands like Kanban (obviously none of the itemtypes are sensitive AFAIK) if they interact with these sensitive itemtypes?

I don't like that at all.

Author: It is possible to implement a response indicating that re-authentication is required, and to implement a modal that allows for re-authentication. So, I don't think this prevents the implementation of an SPA.

In fact, I've implemented it (with an ia), but, because of the risks of breaking a lot of things (because of js changes), we choose not to deliver that code.

each entry is a ``CommonGLPI`` class and throws an ``InvalidArgumentException`` otherwise.

Because the massive action POST data is stored and replayed, the whole selection survives the
prompt: the user verifies once, then the action runs on every selected item.

What this means for a specific massive action (core or plugin):

* Nothing to declare on the action itself. Declaring the itemtype sensitive is enough.
* ``showMassiveActionsSubForm()`` runs through an AJAX call, so it must **not** refuse the
  display just because re-authentication is missing. Use the ``$reauth_needed`` flag, as
  ``MassiveAction`` does for the generic *Update* action:

  .. code-block:: php

     <?php

     // Display the sub-form if the right is granted, or if only a re-authentication
     // is missing (the action will be possible once re-authenticated).
     // No redirection is possible here: the file is called through AJAX.
     $reauth_needed = null;
     $allowed = $item->canGlobal(UPDATE, $reauth_needed);
     if (!$allowed && !$reauth_needed) {
         throw new AccessDeniedHttpException('Missing authorization');
     }

> * ``processMassiveActionsForOneItemtype()`` needs no change: it is only reached after the

Reviewer 3: ```suggestion
* ``processMassiveActionsForOneItemtype()`` needs no change; it is only reached after the
```
